mlflow/xgboost.py

```diff
@@ -416,10 +437,36 @@ def log_feature_importance_plot(features, importance, importance_type):
             finally:
                 shutil.rmtree(tmpdir)
 
-        try_mlflow_log(log_model, model, artifact_path="model")
+        # dtrain must exist as the original train function already ran successfully
+        dtrain = args[1] if len(args) > 1 else kwargs["dtrain"]
+
+        input_example = None
+        signature = None
+        try:
+            input_example = dtrain.data_copy
+
+            # it is possible that the dataset was constructed before the patched
+            #   constructor was applied, so we cannot assume the data_copy exists
+            if input_example is None:
+                raise Exception("failed to gather example input.")
+
+            model_output = model.predict(xgb.DMatrix(input_example))
+            signature = infer_signature(input_example, model_output)
+        except Exception as e:  # pylint: disable=broad-except
+            msg = "Failed to infer the model signature: " + str(e)
+            _logger.warning(msg)
+
+        try_mlflow_log(
+            log_model,
+            model,
+            artifact_path="model",
+            signature=signature,
+            input_example=input_example,
+        )
 
         if auto_end_run:
             try_mlflow_log(mlflow.end_run)
         return model
 
     wrap_patch(xgboost, "train", train)
+    wrap_patch(xgboost.DMatrix, "__init__", __init__)
```

Comment:
Lools like it's impossible to extract given `data` from `xgboost.DMatrix`: https://github.com/dmlc/xgboost/issues/3173